### PR TITLE
Disable loading of workspace configs by default and make it configurable

### LINKED ISSUE #9514: Stop silently executing arbitrary code from the current directory by default
### Summary

According to the [docs about languages](https://docs.helix-editor.com/languages.html#languagestoml-files) language configuration is read, among others, from the current projects .helix directory and allows specification of languages and language server commands. This can easily be exploited to make helix execute arbitrary code when started in an untrusted directory.

Vim disables reading project specific configurations by default and [warns about enabling it](https://vimhelp.org/options.txt.html#%27exrc%27) for this very reason.

:exclamation: PLEASE PATCH THIS AND DON'T USE HELIX IN ANY UNTRUSTED DIRECTORIES UNTIL THEN :exclamation: 

### Reproduction Steps

1. E. g. for a python project create a file *.helix/languages.toml* with the following content.

```toml
[language-server.evil]
command = "sh"
args = ["-c", "echo evil > evil.txt"]

[[language]]
name = "python"
language-servers = ["evil"]
```

2. Start helix with `hx some-file.py`
3. See a file *evil.txt* being automatically created
4. Shiver in fear

### Helix log

_No response_

### Platform

Linux (probably all)

### Terminal Emulator

all

### Installation Method

releases page

### Helix Version

23.10

### PR DESCRIPTION
**Pull request by the-dipsy**
Makes loading of workspace-local `config.toml` and `languages.toml` files optional and disables it by default to protect against vulnerability to arbitrary code execution when running helix in an untrusted directory. Should fix #9514 and alleviate #2697.

### CODE REVIEW

**Review by the-mikedavis — CHANGES_REQUESTED** [review]
I'm not sure if we will want this behavior once we have workspace trust. Instead I think we will want the opposite: the ability to say that you never want to be prompted about workspace trust and never use workspace config. (So the default would be the opposite: that you do merge workspace configs but only when you trust the workspace.)

File: languages.toml [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Language support configuration.
 # See the languages documentation: https://docs.helix-editor.com/master/languages.html
 
+workspace-config = false
```
Comment (archseer) [inline languages.toml:4]: Additionally, `workspace-config` should not be in languages.toml.

**Comment by the-dipsy** [comment]
@the-mikedavis I'd be fine with your workspace trust concept once it is implemented and if the prompt is absolutely clear about the risks. I'd really like the ability to activate language servers and loading of workspace configs separately though.

**Comment by pascalkuthe** [comment]
> I'd really like the ability to activate language servers and loading of workspace configs separately though.

that ssesm like security theater to me.  VSCode doesn't do that either. If you trust the LSP you trust the config. The other way around you can disable LSP in your workspace config.

**Comment by the-dipsy** [comment]
@archseer There is one `workspace-config = true` for `config.toml` and one for `languages.toml`. The current code structure makes it way easier to implement this way and I think it makes sense too :shrug:

File: helix-term/src/config.rs [hunk 2]
```diff
@@ -57,72 +60,45 @@ impl Display for ConfigLoadError {
 
 impl Config {
     pub fn load(
-        global: Result<String, ConfigLoadError>,
-        local: Result<String, ConfigLoadError>,
+        mut global: ConfigRaw, mut workspace: Option<ConfigRaw>,
```
[the remark this answers is not in the document]
Comment (the-dipsy) [reply helix-term/src/config.rs:63]: I wanted to change as little as possible but I'm all for cleaning stuff up a little. Is `Config::load_default` even the right name for what it's doing? I would expect `load_default` to only load the config shipped with the package or at least only the global config.

File: helix-loader/src/config.rs [hunk 3]
```diff
@@ -7,40 +8,28 @@ pub fn default_lang_config() -> toml::Value {
...
-        //   [[language]]
-        //   name = "toml"
-        //   language-server = { command = "/usr/bin/taplo" }
-        //
-        // thus it overrides the third depth-level of b with values of a if they exist, but otherwise merges their values
-        crate::merge_toml_values(a, b, 3)
-    });
+    let global = merge_language_config(default_lang_config(), crate::lang_config_file())?;
+
+    let config = match global.get("workspace-config").and_then(|v| v.as_bool()) {
```
[the remark this answers is not in the document]
Comment (the-dipsy) [reply helix-loader/src/config.rs:29]: How would you implement that with the current code structure? Loading the entire config a second time seems wasteful but the current control flow doesn't seem to make it trivial to pass a parameter loaded from the `config.toml` to the code that loads the `languages.toml`.

**Comment by the-dipsy** [comment]
@pascalkuthe It seems entirely reasonable to me to have language servers statically analyze code, provide snippets etc. without opening up the gates for ACE :flushed: I'm not a heavy user of language servers but this is what GPT-4 says about it:

**User:** How common is unprotected execution of code from currently edited code repositories in language servers used in IDEs?
**GPT-4:** Unprotected execution of code from currently edited repositories in language servers is not common and is generally considered a security risk. Language servers typically provide features like code analysis, linting, autocomplete, and refactoring. They parse the code and understand its structure, but they do not execute the actual code being edited. When execution is necessary, for example, during debugging sessions, it is done in a controlled environment, often with explicit user consent and understanding of the risks.

I'd assume that a lot of people who switch to a small-code-base editor that doesn't rely on plugins, advertises to be usable over SSH, and is implemented in a security-driven language expect it to follow a security-conscious approach and configurability.

**Comment by pascalkuthe** [comment]
gpt is not a reliable source for *anything* and as usual provides a wordsoup without substance.

The majority of languageserver will execute arbitrary code which is why vscode disables them by default. For example rust-analyzer will execute all "build.rs" files and procedural macros (yes without any sandboxing). They often even have their own config files which they read from the project from which they can and do execute commands.

**Comment by the-dipsy** [comment]
@pascalkuthe Certainly no proof and unreliable but often decent at reiterating and summarizing information. I know that rust does this but it therefore is also pretty much the first thing that comes up when putting "language server" and "arbitrary code execution" into a search engine. I don't have a reliable source that says *most* LSPs don't do this. Do you have a source for the opposite? Jedi [apparently is one explicit example](https://jedi.readthedocs.io/en/latest/docs/features.html#for-script). In any case, the fact that there are plenty of ways to use the language server protocol without arbitrary code execution warrants an option to use them that way IMO.

File: helix-term/src/config.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (the-mikedavis) [reply helix-term/src/config.rs:63]: I believe these changes will make workspace trust easier to add so it seems like a worthwhile refactor anyways. Yeah I think we could rename `load_default` as `load` and that would make more sense

File: helix-loader/src/config.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (the-mikedavis) [reply helix-loader/src/config.rs:29]: I think the only option is passing down a boolean parameter. IIRC we always load normal config before language config so it should be possible to read this only from `config.toml`

**Review by the-mikedavis — COMMENTED** [review]
I'm not keen to accept this as-is: we usually do not accept "bandaid" fixes that work around missing features, especially when we would have to remove them or make breaking changes to them later as we would with workspace trust. IMO this is not a very serious threat and seems mostly theoretical so it would be an overreaction to hot-fix it that leaves us with annoying defaults, a lot like git's `safe.directory` response to [cve-2022-24765](https://github.blog/2022-04-12-git-security-vulnerability-announced/#cve-2022-24765) (reiterating what I was saying [here](https://github.com/helix-editor/helix/pull/3207#issuecomment-1257309846)).

But I think this PR could be a really good base for the changes we need for workspace trust and could be merged before that is introduced with a small change: let's switch the `workspace_config: bool` option out for an enum

```rust
#[derive(Default)]
pub enum LoadWorkspaceConfig {
	Always,
    #[default]
    Never,
}
```

which you would write in config as `"always"` or `"never"`. Workspace trust would add a `Trust` variant that should become the default. That way you can decide what behavior you want: always or never load workspace configs, and the meaning of these options doesn't change at all once workspace trust is added.

**Comment by archseer** [comment]
I pushed my set of changes based on some comments above here: https://github.com/helix-editor/helix/commit/5ed223f9476810a65c427c464d99950f1df6ec49

**Comment by the-dipsy** [comment]
@archseer Oh, shoot. I refactored a bunch already. My config merging is not particularly clean yet though.

**Comment by the-dipsy** [comment]
@archseer Oh, shoot. I refactored a bunch already. Is my version okay for you?

File: helix-term/src/config.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (the-dipsy) [reply helix-term/src/config.rs]: Should be done now.

**Comment by LennyPenny** [comment]
@the-dipsy what happened to this PR? Is possible to configure helix to be secure by default now in some other way?

**Comment by the-dipsy** [comment]
> @the-dipsy what happened to this PR? Is possible to configure helix to be secure by default now in some other way? 

I didn't continue adapting my solution to the requirements by the core devs and thus it was never merged. I haven't followed the progress of the workspace trust feature envisioned by them. If it hasn't been finished yet, this ACE vulnerability probably still persists.

**Pull request closed without merging**